### ISSUE TITLE
Unable to build CPU only binaries on macOS

### ISSUE THREAD

**Issue by thomthom** [PixarAnimationStudios/OpenSubdiv#1236]
I have a project where I'm only using osd_cpu.

I configure the build using:

```
mkdir build
cd build
cmake \
  -D NO_EXAMPLES=1 \
  -D NO_TUTORIALS=1 \
  -D NO_REGRESSION=1 \
  -D NO_PTEX=1 \
  -D NO_DOC=1 \
  -D NO_OMP=1 \
  -D NO_TBB=1 \
  -D NO_CUDA=1 \
  -D NO_OPENCL=1 \
  -D NO_CLEW=1 \
  -D NO_OPENGL=1 \
  -D NO_METAL=1 \
  -D NO_DX=1 \
  -D NO_TESTS=1 \
  -D NO_GLTESTS=1 \
  -D NO_GLEW=1 \
  -D NO_GLFW=1 \
  -D NO_GLFW_X11=1 \
  ..
```

This fails to configure with the warning with the output:

```
-- The C compiler identification is AppleClang 12.0.5.12050022
-- The CXX compiler identification is AppleClang 12.0.5.12050022
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Deprecation Warning at CMakeLists.txt:27 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Compiling OpenSubdiv version v3_4_4
-- Using cmake version 3.21.3
-- Configuring done
CMake Error at opensubdiv/CMakeLists.txt:320 (add_library):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:osd_gpu_obj>

  Objects of target "osd_gpu_obj" referenced but no such target exists.


CMake Error at opensubdiv/CMakeLists.txt:273 (add_library):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:osd_gpu_obj>

  Objects of target "osd_gpu_obj" referenced but no such target exists.


CMake Error at opensubdiv/CMakeLists.txt:320 (add_library):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:osd_gpu_obj>

  Objects of target "osd_gpu_obj" referenced but no such target exists.


CMake Error at opensubdiv/CMakeLists.txt:273 (add_library):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:osd_gpu_obj>

  Objects of target "osd_gpu_obj" referenced but no such target exists.


CMake Error at opensubdiv/CMakeLists.txt:320 (add_library):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:osd_gpu_obj>

  Objects of target "osd_gpu_obj" referenced but no such target exists.


CMake Error at opensubdiv/CMakeLists.txt:273 (add_library):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:osd_gpu_obj>

  Objects of target "osd_gpu_obj" referenced but no such target exists.


CMake Error at opensubdiv/CMakeLists.txt:320 (add_library):
  No SOURCES given to target: osd_dynamic_framework


CMake Error at opensubdiv/CMakeLists.txt:273 (add_library):
  No SOURCES given to target: osd_static_framework


CMake Generate step failed.  Build files cannot be regenerated correctly.
```

The errors are coming from the frameworks being build on the `APPLE` platform:

https://github.com/PixarAnimationStudios/OpenSubdiv/blob/bb0b366cb7591411b0bac0291f41794751461392/opensubdiv/CMakeLists.txt#L272-L282

https://github.com/PixarAnimationStudios/OpenSubdiv/blob/bb0b366cb7591411b0bac0291f41794751461392/opensubdiv/CMakeLists.txt#L320-L329

They are blindly including the `osd_gpu_obj` target even though that target will not be generated with the above config:

https://github.com/PixarAnimationStudios/OpenSubdiv/blob/bb0b366cb7591411b0bac0291f41794751461392/opensubdiv/osd/CMakeLists.txt#L387-L399

Atm this is blocking me. I need to patch it locally for it to work for me. Question is, why is the framework always including the gpu target?

The static and dynamic library targets doesn't do this:

https://github.com/PixarAnimationStudios/OpenSubdiv/blob/bb0b366cb7591411b0bac0291f41794751461392/opensubdiv/CMakeLists.txt#L127-L135

https://github.com/PixarAnimationStudios/OpenSubdiv/blob/bb0b366cb7591411b0bac0291f41794751461392/opensubdiv/CMakeLists.txt#L173-L183

**Comment by thomthom** [comment]
Also, would it not be useful to have an option to disable the build of the frameworks? I'm using static libs, and have no need for the frameworks generated.

**Comment by thomthom** [comment]
The framework configuration appears to have been added as part of Metal support: https://github.com/PixarAnimationStudios/OpenSubdiv/commit/30cd2df

Though a `NO_METAL` option was added the frameworks targets seems to assume that there is always some GPU target available. (Maybe a CI test with CPU only config would be useful?)

**Comment by davidgyu** [comment]
Thanks @thomthom These are good suggestions.
You mentioned being blocked. Have you been able to workaround this locally?

**Comment by thomthom** [comment]
Yea, I created a patch branch on my fork that simply disabled generation of the .frameworks. I couldn't figure out a quick-fix for generating frameworks without GPU or adding option for not generating frameworks.

https://github.com/thomthom/OpenSubdiv/commit/8c27441f0ec5e4fc64d8992ed667ea3dc323e31f

**Comment by thomthom** [comment]
I'm wondering, the generation of the frameworks was added for Metal support. Was that a requirement for Metal?

If an option to turn off generation of frameworks was added, would that affect the Metal build in any way? I'm trying to figure out if option for framework and the framework  being able to build CPU only is two separate issues or somehow interlinked.

**Comment by davidgyu** [comment]
The framework bundles are not required for the Metal implementation to work correctly. So, an option seems reasonable.

**Comment by jilliene** [comment]
Filed as internal issue #OSD-362

**Comment by ferdnyc** [comment]
@thomthom

> They are blindly including the osd_gpu_obj target

```cmake
 #static framework 
 add_library(osd_static_framework 
     STATIC 
     version.cpp 
     $<TARGET_OBJECTS:sdc_obj> 
     $<TARGET_OBJECTS:vtr_obj> 
     $<TARGET_OBJECTS:far_obj> 
     $<TARGET_OBJECTS:osd_cpu_obj> 
     $<TARGET_OBJECTS:osd_gpu_obj> 
     ${OPENGL_LOADER_OBJS} 
 ) 
```

CMake has a fix for that! Since 3.12, anyway. If being able to build the frameworks would be useful to anyone, even without the GPU code. 

The solution has two parts:

1. Don't use `$<TARGET_OBJECTS:>` for `OBJECT` libraries (redundant).
2. DO use `$<TARGET_NAME_IF_EXISTS>` for optional targets

If a library is an `OBJECT` library, `$<TARGET_OBJECTS>` is implicit. So you can use `target_link_libraries` to incorporate it into other libs, instead:

```cmake
 add_library(osd_static_framework 
     STATIC 
     version.cpp 
    ${OPENGL_LOADER_OBJS} 
)
target_link_libraries(osd_static_framework PRIVATE
    sdc_obj
    vtr_obj
    far_obj
    osd_cpu_obj
    osd_gpu_obj
 ) 
```

(It might **appear** that CMake is doing more work with things set up this way, as it adds some verbosity to the build process, but it's just extra chatter. You can look in the build dir and see, the different targets' output dirs (`${PROJECT_BINARY_DIR}/opensubdiv/CMakeFiles/*.dir`) have no duplicates in them except for the `version.cpp.o` objects.)

And now that you're linking objects via the targets themselves, rather than using their object files as sources, you can wrap any optional `OBJECT` targets in [`$<TARGET_NAME_IF_EXISTS>](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html?highlight=genex_eval#genex:TARGET_NAME_IF_EXISTS) so it won't matter if it's missing:

```cmake
 add_library(osd_static_framework 
     STATIC 
     version.cpp 
    ${OPENGL_LOADER_OBJS} 
)
target_link_libraries(osd_static_framework PRIVATE
    sdc_obj
    vtr_obj
    far_obj
    osd_cpu_obj
    $<TARGET_NAME_IF_EXISTS:osd_gpu_obj>
 ) 
```

(Probably the way `${OPENGL_LOADER_LIBS}` is incorporated should also be changed, but you get the idea.)

**Comment by davidgyu** [comment]
Thank you for these notes @ferdnyc
